Modules/Core/Common/include/itkIndexRange.h

```diff
@@ -0,0 +1,429 @@
+    m_MinIndex(imageRegion.GetIndex()),
+    m_MaxIndex(CalculateMaxIndex(imageRegion.GetIndex(), imageRegion.GetSize()))
+  {
+    // Three compile-time asserts, just to check if SFINAE worked properly:
+    static_assert(!VIsSubstitutionFailure,
+      "This template should (of course) be instantiated without substitution failure.");
+    static_assert(std::is_same<TVoid, void>::value,
+      "std::enable_if<!VIsSubstitutionFailure> should yield void, by definition.");
+    static_assert(!VBeginAtZero,
+      "This constructor should only be is available when VBeginAtZero is false.");
+  }
+
+
+  /** Returns an iterator to the first index. */
+  iterator begin() const ITK_NOEXCEPT
+  {
+    return iterator(m_MinIndex, m_MinIndex, m_MaxIndex);
+  }
+
+  /** Returns an 'end iterator' for this range. */
+  iterator end() const ITK_NOEXCEPT
+  {
+    IndexType index = m_MinIndex;
+    index.back() = m_MaxIndex.back() + 1;
+    return iterator(index, m_MinIndex, m_MaxIndex);
+  }
+
+  /** Returns a const iterator to the first index.
+    * Provides only read-only access to the index data. */
+  const_iterator cbegin() const ITK_NOEXCEPT
+  {
+    return this->begin();
+  }
+
+  /** Returns a const 'end iterator' for this range. */
+  const_iterator cend() const ITK_NOEXCEPT
+  {
+    return this->end();
+  }
+
+  /** Returns a reverse 'begin iterator' for this range. */
+  reverse_iterator rbegin() const ITK_NOEXCEPT
+  {
+    return reverse_iterator(this->end());
+  }
+
+  /** Returns a reverse 'end iterator' for this range. */
+  reverse_iterator rend() const ITK_NOEXCEPT
+  {
+    return reverse_iterator(this->begin());
+  }
+
+  /** Returns a const reverse 'begin iterator' for this range. */
+  const_reverse_iterator crbegin() const ITK_NOEXCEPT
+  {
+    return this->rbegin();
+  }
+
+  /** Returns a const reverse 'end iterator' for this range. */
+  const_reverse_iterator crend() const ITK_NOEXCEPT
+  {
+    return this->rend();
+  }
+
+
+  /** Returns the size of the range, that is the number of indices. */
+  std::size_t size() const ITK_NOEXCEPT
+  {
+    std::size_t result = 1;
+
+    for (unsigned i = 0; i < VDimension; ++i)
+    {
+      result *= ((m_MaxIndex[i] + 1) - m_MinIndex[i]);
+    }
+    return result;
+  }
+
+private:
+
+  using MinIndexType = typename iterator::MinIndexType;
+
+  static IndexType CalculateMaxIndex(const MinIndexType& minIndex, const SizeType& size)
+  {
+    IndexType index;
+
+    for (unsigned i = 0; i < VDimension; ++i)
+    {
+      index[i] = minIndex[i] + static_cast<IndexValueType>(size[i]) - 1;
+    }
+
+    return index;
+  }
+
+  // IndexRange data members:
+
+  // Minimum (N-dimensional) index.
+  MinIndexType m_MinIndex;
+
+  // Maximum (N-dimensional) index.
+  IndexType m_MaxIndex;
+
+};
+
+template<unsigned VDimension>
+using ImageRegionIndexRange = IndexRange<VDimension, false>;
+
+template<unsigned VDimension>
+using ZeroBasedIndexRange = IndexRange<VDimension, true>;
```

Comment:
Update: the latest amend has an extra type alias, `ZeroBasedIndexRange<Dimension>`, optimized for zero-based index iteration.

Reply:
Sweet! :candy: